images/cfw/opt/x1plusd.py

```diff
@@ -1,116 +1,282 @@
+                method = msg.header.fields[HeaderFields.member]
+
+                if msg.header.fields[HeaderFields.signature] != "s":
+                    await self.router.send(
+                        new_error(msg, "x1plus.x1plusd.Error.BadSignature")
+                    )
+                    continue
+
+                arg = msg.body[0]
+
+                rv = None
+                try:
+                    if method == "ota":
+                        rv = await self._update_check(json.loads(arg))
+
+                except Exception as e:
+                    logger.error(f"{method}({arg}) -> exception {e}")
+                    await self.router.send(
+                        new_error(msg, "x1plus.x1plusd.Error.InternalError")
+                    )
+                    continue
+
+                if not rv:
+                    logger.warning(f"{method} -> NoMethod")
+                    await self.router.send(
+                        new_error(msg, "x1plus.x1plusd.Error.NoMethod")
+                    )
+                    continue
+
+                logger.debug(f"{method}({arg}) -> {rv}")
+                await self.router.send(new_method_return(msg, "s", (rv,)))
+
+    async def _update_check(self, payload):
+        ota_response = {
+            "ota_available": self.ota_available,
+            "err_on_last_check": self.last_check_error,
+            "last_checked": self.last_check_timestamp,
+            "ota_info": self.last_check_response,
+            "is_downloaded": self.ota_downloaded,
+        }
+
+        # First, if we were asked just for the last status, just return our ota object
+        if not payload.get("check", False):
+            return json.dumps({"ota": ota_response})
+
+        # Load in our settings.json, we need to make sure we are still enabled! :)
+        try:
+            with open(
+                f"/mnt/sdcard/x1plus/printers/{_get_sn()}/settings.json", "r"
+            ) as fh:
+                x1p_settings = json.load(fh)
+        except FileNotFoundError:
+            logger.warning("OTA can't find settings.json, assuming OTAs are disabled!")
+            x1p_settings = {"ota.enable": False}
+
+        # Do we have OTAs enabled? If not, just return current status
+        if not x1p_settings.get("ota.enable", False):
+            logger.info("OTA check is disabled, skipping check!")
+            return json.dumps({"ota": ota_response})
+
+        # If we are here we want to check, so check
+        try:
+            # Update check timestamp first
+            self.last_check_timestamp = datetime.datetime.now().timestamp()
+            r = requests.get(self.ota_url, timeout=5)
+            self.last_check_response = r.json()
+        except Exception as e:
+            logger.error(f"Exception calling OTA URL! Error of: {e}")
+            # we Timed out, or hit other error with requests
+            self.last_check_error = True
+            return json.dumps({"ota": ota_response})
```

Comment:
We shouldn't do anything synchronous on the Dbus thread -- that will block every other Dbus user in x1plusd.  It's ok to use aiohttp here, but not asyncio.  It's also okay to have a background thread. 

There's also another problem, though, which is that a Dbus API call should be 'fast', because it blocks the UI thread for bbl_screen.  Maybe the API is, like:

```
x1plus.ota.GetStatus({}) -> {status structure}
x1plus.ota.CheckNow({}) -> {} # starts the ota background task checking, returns immediately
```

And then as things change in the X1Plus OTA world, it emits signals (the way the settings service does):
```
x1plus.ota.Status({status structure})
```

And the status structure is something like:
```javascript
{
  "check_enabled": ...,
  "check_interval": ...
  "check_timestamp": timestamp,
  "check_status": "OK" /* or "CHECKING", or "CHECK_FAILED" */,
  "version_ota": { version info... },
  "version_current": { version info... },
  "download_status": { some structure? },
}
```

or something?

Reply:
hmm, im not againt this, because you are right about async for the request. Any chance we can get aiohttp or grequests added to the python 3.12 PR and that becomes a merge dependency before OTAs?

Reply:
I was going to say that I think this can be orthogonal to 3.12, but it turns out that there are native components for aiohttp.  So it would be easier to do this simultaneously with 3.12, I guess.  Anyway, now that is on the jwise/python3.12 branch, feel free to check it out.

I wonder if we should build python cache files at install time / packaging time.  First startup is insanely slow.